src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -299,6 +308,65 @@ impl LogicalJoin {
             .unwrap_or_default()
     }
 
+    pub(super) fn derive_fd(
+        left_len: usize,
+        right_len: usize,
+        left_fd_set: FunctionalDependencySet,
+        right_fd_set: FunctionalDependencySet,
+        on: &Condition,
+        join_type: JoinType,
+        output_indices: &[usize],
+    ) -> FunctionalDependencySet {
+        let out_col_num = Self::out_column_num(left_len, right_len, join_type);
+
+        let get_new_left_fd_set = |left_fd_set: FunctionalDependencySet| {
+            ColIndexMapping::with_shift_offset(left_len, 0)
+                .composite(&ColIndexMapping::identity(out_col_num))
+                .rewrite_functional_dependency_set(left_fd_set)
+        };
+        let get_new_right_fd_set = |right_fd_set: FunctionalDependencySet| {
+            ColIndexMapping::with_shift_offset(right_len, left_len.try_into().unwrap())
+                .rewrite_functional_dependency_set(right_fd_set)
+        };
+        let fd_set: FunctionalDependencySet = match join_type {
+            JoinType::Inner => {
+                let mut fd_set = FunctionalDependencySet::new(out_col_num);
+                for i in &on.conjunctions {
+                    if let Some((col, _)) = i.as_eq_const() {
+                        fd_set.add_constant_columns(&[col.index()])
+                    } else if let Some((left, right)) = i.as_eq_cond() {
+                        fd_set.add_functional_dependency_by_column_indices(
+                            &[left.index()],
+                            &[right.index()],
+                        );
+                        fd_set.add_functional_dependency_by_column_indices(
+                            &[right.index()],
+                            &[left.index()],
+                        );
+                    }
+                }
+                get_new_left_fd_set(left_fd_set)
+                    .into_dependencies()
+                    .into_iter()
+                    .chain(
+                        get_new_right_fd_set(right_fd_set)
+                            .into_dependencies()
+                            .into_iter(),
+                    )
+                    .for_each(|fd| fd_set.add_functional_dependency(fd));
+                fd_set
+            }
+            JoinType::LeftOuter => get_new_right_fd_set(right_fd_set),
+            JoinType::RightOuter => get_new_left_fd_set(left_fd_set),
```

Comment:
LOJ should inherit from left? Because left is unchanged, while right is padding with NULLs